graphite/graphite.go

```diff
@@ -0,0 +1,330 @@
+// committer goroutine.
+func (s *Server) handleMessage(r *bufio.Reader) error {
+	// Decode graphic metric.
+	m, err := DecodeMetric(r)
+	if err != nil {
+		return err
+	}
+
+	// Convert metric to a field value.
+	var values = make(map[string]interface{})
+	if m.IsInt {
+		values[m.Name] = &m.IntegerValue
+	} else {
+		values[m.Name] = &m.FloatValue
+	}
+
+	retentionPolicy, err := s.Server.DefaultRetentionPolicy(s.Database)
+
+	if err != nil {
+		return fmt.Errorf("error looking up default database retention policy: %s", err)
+	}
+
+	if err := s.Server.WriteSeries(
+		s.Database,
+		retentionPolicy.Name,
+		m.Name,
+		m.Tags,
+		m.Timestamp,
+		values,
+	); err != nil {
+		return fmt.Errorf("write series data: %s", err)
+	}
+	return nil
+}
+
+type Metric struct {
+	Name         string
+	Tags         map[string]string
+	IsInt        bool
+	IntegerValue int64
+	FloatValue   float64
+	Timestamp    time.Time
+}
+
+// returns err == io.EOF when we hit EOF without any further data
+func DecodeMetric(r *bufio.Reader) (*Metric, error) {
+	// Read up to the next newline.
+	buf, err := r.ReadBytes('\n')
+	if err != nil && err != io.EOF {
+		// it's possible to get EOF but also data
+		return nil, fmt.Errorf("connection closed uncleanly/broken: %s\n", err.Error())
+	}
+	// Trim the buffer, even though there should be no padding
+	str := strings.TrimSpace(string(buf))
+	// Remove line return
+	str = strings.TrimSuffix(str, `\n`)
+	if str == "" {
+		return nil, err
+	}
+	// Break into 3 fields (name, value, timestamp).
+	fields := strings.Fields(str)
+	if len(fields) != 3 {
+		return nil, fmt.Errorf("received %q which doesn't have three fields", str)
+	}
+
+	m := new(Metric)
+	// decode the name and tags
+	name, tags, err := DecodeNameAndTags(fields[0])
+	if err != nil {
+		return nil, err
+	}
+	m.Name = name
+	m.Tags = tags
+
+	// Parse value.
+	v, err := strconv.ParseFloat(fields[1], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine if value is a float or an int.
+	if i := int64(v); float64(i) == v {
+		m.IntegerValue, m.IsInt = int64(v), true
+	} else {
+		m.FloatValue = v
+	}
+
+	// Parse timestamp.
+	unixTime, err := strconv.ParseInt(fields[2], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	m.Timestamp = time.Unix(0, unixTime*int64(time.Millisecond))
+
+	return m, nil
+}
+
+func DecodeNameAndTags(field string) (string, map[string]string, error) {
+	var (
+		name string
+		tags = make(map[string]string)
+	)
+
+	// decode the name and tags
+	values := strings.Split(field, `.`)
+	if len(values)%2 != 1 {
+		// There should always be an odd number of fields to map a metric name and tags
+		// ex: region.us-west.hostname.server01.cpu -> tags -> region: us-west, hostname: server01, metric name -> cpu
+		return name, tags, fmt.Errorf("received %q which doesn't conform to format of key.value.key.value.metric or metric", field)
+	}
+
+	// Name is the last field
+	name = values[len(values)-1]
```

Comment:
They should be able to set `name` as either the first or last field via the config.